src/iccpd/src/mlacp_link_handler.c

```diff
@@ -0,0 +1,2264 @@
+
+    /*If peer link up, set all the mac that point to the peer-link in ASIC*/
+    TAILQ_FOREACH(msg, &MLACP(csm).mac_list, tail)
+    {
+        mac_msg = (struct MACMsg*)msg->buf;
+
+        /* Find the MAC that the port is peer-link to be added*/
+        if (strcmp(mac_msg->ifname, csm->peer_itf_name) != 0)
+            continue;
+
+        ICCPD_LOG_DEBUG(__FUNCTION__, "Peer link up, add MAC to ASIC for peer-link: %s, MAC %s vlan-id %d",
+                        mac_msg->ifname, mac_msg->mac_str, mac_msg->vid);
+
+        /*Send mac add message to mclagsyncd, local age flag is already set*/
+        add_mac_to_chip(mac_msg, MAC_TYPE_DYNAMIC);
+    }
+
+    return;
+}
+
+void mlacp_peerlink_down_handler(struct CSM* csm)
+{
+    struct Msg* msg = NULL;
+    struct MACMsg* mac_msg = NULL;
+
+    if (!csm)
+        return;
+
+    /*If peer link down, remove all the mac that point to the peer-link*/
+    TAILQ_FOREACH(msg, &MLACP(csm).mac_list, tail)
+    {
+        mac_msg = (struct MACMsg*)msg->buf;
+
+        /* Find the MAC that the port is peer-link to be deleted*/
+        if (strcmp(mac_msg->ifname, csm->peer_itf_name) != 0)
+            continue;
+
+        mac_msg->age_flag = set_mac_local_age_flag(csm, mac_msg, 1);
+
+        ICCPD_LOG_DEBUG(__FUNCTION__, "Peer link down, del MAC for peer-link: %s, MAC %s vlan-id %d",
+                        mac_msg->ifname, mac_msg->mac_str, mac_msg->vid);
+
+        /*Send mac del message to mclagsyncd*/
+        del_mac_from_chip(mac_msg);
+
+        /*If peer is not age, keep the MAC in mac_list, but ASIC is deleted*/
+        if (mac_msg->age_flag == (MAC_AGE_LOCAL | MAC_AGE_PEER))
+        {
+            /*If local and peer both aged, del the mac*/
+            TAILQ_REMOVE(&(MLACP(csm).mac_list), msg, tail);
+            free(msg->buf);
+            free(msg);
+        }
+    }
+
+    return;
+}
+
+/*****************************************
+* Po add/remove handler
+*
+*****************************************/
+void mlacp_mlag_link_add_handler(struct CSM *csm, struct LocalInterface *lif)
+{
+    if (!csm || !lif)
+        return;
+    if (MLACP(csm).current_state != MLACP_STATE_EXCHANGE)
+        return;
+
+    set_peerlink_mlag_port_isolate(csm, lif, 1);
+
+    return;
+}
+
+void mlacp_mlag_link_del_handler(struct CSM *csm, struct LocalInterface *lif)
+{
+    if (!csm || !lif)
+        return;
+
+    if (MLACP(csm).current_state != MLACP_STATE_EXCHANGE)
+        return;
+
+    set_peerlink_mlag_port_isolate(csm, lif, 0);
+
+    return;
+}
+
+int iccp_connect_syncd()
+{
+    struct System* sys = NULL;
+    int ret = 0;
+    int fd = 0;
+    struct sockaddr_in serv;
+    static int count = 0;
+    struct epoll_event event;
+
+    if ((sys = system_get_instance()) == NULL)
+        goto conn_fail;
+
+    if (sys->sync_fd >= 0)
+        return 0;
+
+    /*Print the fail log message every 60s*/
+    if (count >= 600)
+    {
+        count = 0;
+    }
+
+    fd = socket(AF_INET, SOCK_STREAM, 0);
```

Comment:
SOCKET_STREAM socket with binding to 127.0.0.6:2626  is used to communicate between mclagsyncd and iccpd process. The way that this TCP socket is used through out the iccp code makes it possible that multiple TCP messages will be aggregated on a single read and only first one get processed. Also not sure if using network socket inside Sonic instance is a recommended approach.

Reply:
A reliable connection is needed here, so SOCKET_STREAM socket is chosen. And we will refine the problem you mentioned soon.